File: packages/remote-bridge/internal/bridge/session_pointers.go
package bridge

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync"
)

type MemorySessionPointers struct {
	mu          sync.Mutex
	sessions    map[string]string
	parents     map[string]string
	eventCursor string
}

func NewMemorySessionPointers() *MemorySessionPointers {
	return &MemorySessionPointers{
		sessions: make(map[string]string),
		parents:  make(map[string]string),
	}
}

func (p *MemorySessionPointers) Get(remoteKey string) string {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.sessions[remoteKey]
}

func (p *MemorySessionPointers) Set(remoteKey string, sessionID string) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.hasRootConflictLocked(remoteKey, sessionID) {
		return fmt.Errorf("session root is already bound to another remote conversation")
	}
	p.sessions[remoteKey] = sessionID
	return nil
}

func (p *MemorySessionPointers) SetParent(sessionID string, parentID string) error {
	if sessionID == "" || parentID == "" {
		return nil
	}
	p.mu.Lock()
	defer p.mu.Unlock()
	if hasAnyRootConflict(p.sessions, withParent(p.parents, sessionID, parentID)) {
		return fmt.Errorf("session root is already bound to another remote conversation")
	}
	p.parents[sessionID] = parentID
	return nil
}

func (p *MemorySessionPointers) RemoteKeyForSession(sessionID string) string {
	p.mu.Lock()
	defer p.mu.Unlock()
	keys := p.remoteKeysForRootLocked(p.rootLocked(sessionID))
	if len(keys) != 1 {
		return ""
	}
	return keys[0]
}

func (p *MemorySessionPointers) RootSession(sessionID string) string {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.rootLocked(sessionID)
}

func (p *MemorySessionPointers) EventCursor() string {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.eventCursor
}

func (p *MemorySessionPointers) SetEventCursor(cursor string) error {
	if cursor == "" {
		return nil
	}
	p.mu.Lock()
	defer p.mu.Unlock()
	p.eventCursor = cursor
	return nil
}

func (p *MemorySessionPointers) rootLocked(sessionID string) string {
	return rootSession(p.parents, sessionID)
}

func (p *MemorySessionPointers) hasRootConflictLocked(remoteKey string, sessionID string) bool {
	return hasRootConflict(p.sessions, p.parents, remoteKey, sessionID)
}

func (p *MemorySessionPointers) remoteKeysForRootLocked(root string) []string {
	return remoteKeysForRoot(p.sessions, p.parents, root)
}

type FileSessionPointers struct {
	mu          sync.Mutex
	path        string
	sessions    map[string]string
	parents     map[string]string
	eventCursor string
}

func NewFileSessionPointers(path string) (*FileSessionPointers, error) {
	pointers := &FileSessionPointers{
		path:     path,
		sessions: make(map[string]string),
		parents:  make(map[string]string),
	}
	data, err := os.ReadFile(path)
	if errors.Is(err, os.ErrNotExist) {
		return pointers, nil
	}
	if err != nil {
		return nil, err
	}
	if len(data) == 0 {
		return pointers, nil
	}
	var state struct {
		Sessions    map[string]string `json:"sessions"`
		Parents     map[string]string `json:"parents"`
		EventCursor string            `json:"eventCursor"`
	}
	if err := json.Unmarshal(data, &state); err == nil && (state.Sessions != nil || state.Parents != nil || state.EventCursor != "") {
		if state.Sessions != nil {
			pointers.sessions = state.Sessions
		}
		if state.Parents != nil {
			pointers.parents = state.Parents
		}
		pointers.eventCursor = state.EventCursor
		return pointers, nil
	}
	if err := json.Unmarshal(data, &pointers.sessions); err != nil {
		return nil, err
	}
	return pointers, nil
}

func (p *FileSessionPointers) Get(remoteKey string) string {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.sessions[remoteKey]
}

func (p *FileSessionPointers) Set(remoteKey string, sessionID string) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.hasRootConflictLocked(remoteKey, sessionID) {
		return fmt.Errorf("session root is already bound to another remote conversation")
	}
	p.sessions[remoteKey] = sessionID
	return p.saveLocked()
}

func (p *FileSessionPointers) SetParent(sessionID string, parentID string) error {
	if sessionID == "" || parentID == "" {
		return nil
	}
	p.mu.Lock()
	defer p.mu.Unlock()
	if hasAnyRootConflict(p.sessions, withParent(p.parents, sessionID, parentID)) {
		return fmt.Errorf("session root is already bound to another remote conversation")
	}
	p.parents[sessionID] = parentID
	return p.saveLocked()
}

func (p *FileSessionPointers) RemoteKeyForSession(sessionID string) string {
	p.mu.Lock()
	defer p.mu.Unlock()
	keys := p.remoteKeysForRootLocked(p.rootLocked(sessionID))
	if len(keys) != 1 {
		return ""
	}
	return keys[0]
}

func (p *FileSessionPointers) RootSession(sessionID string) string {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.rootLocked(sessionID)
}

func (p *FileSessionPointers) EventCursor() string {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.eventCursor
}

func (p *FileSessionPointers) SetEventCursor(cursor string) error {
	if cursor == "" {
		return nil
	}
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.eventCursor == cursor {
		return nil
	}
	p.eventCursor = cursor
	return p.saveLocked()
}

func (p *FileSessionPointers) rootLocked(sessionID string) string {
	return rootSession(p.parents, sessionID)
}

func (p *FileSessionPointers) hasRootConflictLocked(remoteKey string, sessionID string) bool {
	return hasRootConflict(p.sessions, p.parents, remoteKey, sessionID)
}

func (p *FileSessionPointers) remoteKeysForRootLocked(root string) []string {
	return remoteKeysForRoot(p.sessions, p.parents, root)
}

func rootSession(parents map[string]string, sessionID string) string {
	if sessionID == "" {
		return ""
	}
	seen := map[string]bool{}
	current := sessionID
	for current != "" && !seen[current] {
		seen[current] = true
		parent := parents[current]
		if parent == "" {
			return current
		}
		current = parent
	}
	return sessionID
}

func hasRootConflict(sessions map[string]string, parents map[string]string, remoteKey string, sessionID string) bool {
	root := rootSession(parents, sessionID)
	if root == "" {
		return false
	}
	for currentKey, currentSession := range sessions {
		if currentKey != remoteKey && rootSession(parents, currentSession) == root {
			return true
		}
	}
	return false
}

func hasAnyRootConflict(sessions map[string]string, parents map[string]string) bool {
	seen := map[string]string{}
	for remoteKey, sessionID := range sessions {
		root := rootSession(parents, sessionID)
		if root == "" {
			continue
		}
		if current := seen[root]; current != "" && current != remoteKey {
			return true
		}
		seen[root] = remoteKey
	}
	return false
}

func withParent(parents map[string]string, sessionID string, parentID string) map[string]string {
	next := make(map[string]string, len(parents)+1)
	for key, value := range parents {
		next[key] = value
	}
	next[sessionID] = parentID
	return next
}

func remoteKeysForRoot(sessions map[string]string, parents map[string]string, root string) []string {
	if root == "" {
		return nil
	}
	var keys []string
	for remoteKey, current := range sessions {
		if rootSession(parents, current) == root {
			keys = append(keys, remoteKey)
		}
	}
	return keys
}

func (p *FileSessionPointers) saveLocked() error {
	if err := os.MkdirAll(filepath.Dir(p.path), 0o700); err != nil {
		return err
	}
	data, err := json.MarshalIndent(struct {
		Sessions    map[string]string `json:"sessions"`
		Parents     map[string]string `json:"parents"`
		EventCursor string            `json:"eventCursor,omitempty"`
	}{
		Sessions:    p.sessions,
		Parents:     p.parents,
		EventCursor: p.eventCursor,
	}, "", "  ")
	if err != nil {
		return err
	}
	tempPath := p.path + ".tmp"
	if err := os.WriteFile(tempPath, data, 0o600); err != nil {
		return err
	}
	return os.Rename(tempPath, p.path)
}

File: packages/remote-bridge/src/session-pointers.ts
import { mkdir, readFile, rename, writeFile, unlink } from "node:fs/promises"
import { dirname } from "node:path"

// Process-wide so two stores sharing a path (same pid) never collide on a temp
// name or rename over the same state file at the same time.
let tempSequence = 0
const writeQueues = new Map<string, Promise<void>>()

/**
 * Maps remote conversations (a platform-scoped key) to PawWork sessions and
 * tracks the SSE event cursor. With a non-empty path it persists to disk; an
 * empty/undefined path keeps everything in memory.
 *
 * Ported from the Go `bridge.SessionPointersStore`. JS runs the in-memory map
 * mutations on a single thread, so the explicit mutex is gone; disk writes for
 * the same path are still serialized and land atomically through a unique temp
 * file + rename, so an interrupted or interleaved write can never leave a
 * partial snapshot on disk.
 */
export class SessionPointers {
  private readonly path: string
  private sessions = new Map<string, string>()
  private parents = new Map<string, string>()
  private cursor = ""

  constructor(path = "") {
    this.path = path
  }

  static memory(): SessionPointers {
    return new SessionPointers("")
  }

  /** Load a persisted store, or start empty when the file is absent/blank. */
  static async fromFile(path: string): Promise<SessionPointers> {
    const store = new SessionPointers(path)
    let data: string
    try {
      data = await readFile(path, "utf8")
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code === "ENOENT") return store
      throw err
    }
    if (data.trim() === "") return store
    const parsed = JSON.parse(data) as Record<string, unknown>
    const isWrapped = parsed.sessions !== undefined || parsed.parents !== undefined || parsed.eventCursor !== undefined
    if (isWrapped) {
      const state = parsed as { sessions?: Record<string, string>; parents?: Record<string, string>; eventCursor?: string }
      if (state.sessions) store.sessions = new Map(Object.entries(state.sessions))
      if (state.parents) store.parents = new Map(Object.entries(state.parents))
      store.cursor = state.eventCursor ?? ""
    } else {
      // Legacy bare-map format ({"<remoteKey>":"<sessionID>"}) written by an
      // early build. Match the Go loader's fallback so an upgrade keeps its
      // chat-to-session bindings instead of silently starting empty.
      store.sessions = new Map(Object.entries(parsed as Record<string, string>))
    }
    return store
  }

  get(remoteKey: string): string {
    return this.sessions.get(remoteKey) ?? ""
  }

  async set(remoteKey: string, sessionID: string): Promise<void> {
    if (hasRootConflict(this.sessions, this.parents, remoteKey, sessionID)) {
      throw new Error("session root is already bound to another remote conversation")
    }
    this.sessions.set(remoteKey, sessionID)
    await this.save()
  }

  async setParent(sessionID: string, parentID: string): Promise<void> {
    if (sessionID === "" || parentID === "") return
    if (parentChainReaches(this.parents, parentID, sessionID)) {
      throw new Error("session parent would create a cycle")
    }
    if (hasAnyRootConflict(this.sessions, withParent(this.parents, sessionID, parentID))) {
      throw new Error("session root is already bound to another remote conversation")
    }
    this.parents.set(sessionID, parentID)
    await this.save()
  }

  /** The remote key bound to a session's root, or "" when none or ambiguous. */
  remoteKeyForSession(sessionID: string): string {
    const keys = remoteKeysForRoot(this.sessions, this.parents, rootSession(this.parents, sessionID))
    return keys.length === 1 ? keys[0] : ""
  }

  rootSession(sessionID: string): string {
    return rootSession(this.parents, sessionID)
  }

  eventCursor(): string {
    return this.cursor
  }

  async setEventCursor(cursor: string): Promise<void> {
    if (cursor === "" || this.cursor === cursor) return
    this.cursor = cursor
    await this.save()
  }

  private save(): Promise<void> {
    if (this.path === "") return Promise.resolve()
    // Snapshot synchronously so the queued write reflects this call's state.
    const snapshot = JSON.stringify(
      {
        sessions: Object.fromEntries(this.sessions),
        parents: Object.fromEntries(this.parents),
        ...(this.cursor ? { eventCursor: this.cursor } : {}),
      },
      null,
      2,
    )
    return queuePathWrite(this.path, () => this.writeSnapshot(snapshot))
  }

  private async writeSnapshot(snapshot: string): Promise<void> {
    const dir = dirname(this.path)
    await mkdir(dir, { recursive: true, mode: 0o700 })
    // Unique temp name per write so a second writer sharing this path cannot
    // clobber a fixed `<path>.tmp` mid-write; each renames its own snapshot.
    const tempPath = `${this.path}.${process.pid}.${tempSequence++}.tmp`
    try {
      // Mode 0o600 so the state file (remote keys / session IDs / cursor) is not
      // world/group-readable under a shared dir; matches Go's os.CreateTemp default.
      await writeFile(tempPath, snapshot, { mode: 0o600 })
      await rename(tempPath, this.path)
    } catch (err) {
      await unlink(tempPath).catch(() => {})
      throw err
    }
  }
}

/** Walk the parent chain from sessionID to its root; cycle-safe. */
export function rootSession(parents: Map<string, string>, sessionID: string): string {
  if (sessionID === "") return ""
  const seen = new Set<string>()
  let current = sessionID
  while (current !== "" && !seen.has(current)) {
    seen.add(current)
    const parent = parents.get(current) ?? ""
    if (parent === "") return current
    current = parent
  }
  return sessionID
}

/**
 * Whether walking the parent chain from `start` ever lands on `target`.
 * setParent uses it to reject a parentID whose ancestry already contains the
 * child, which would otherwise form a cycle.
 */
function parentChainReaches(parents: Map<string, string>, start: string, target: string): boolean {
  const seen = new Set<string>()
  for (let current = start; current !== "" && !seen.has(current); current = parents.get(current) ?? "") {
    if (current === target) return true
    seen.add(current)
  }
  return false
}

function hasRootConflict(
  sessions: Map<string, string>,
  parents: Map<string, string>,
  remoteKey: string,
  sessionID: string,
): boolean {
  const root = rootSession(parents, sessionID)
  if (root === "") return false
  for (const [currentKey, currentSession] of sessions) {
    if (currentKey !== remoteKey && rootSession(parents, currentSession) === root) return true
  }
  return false
}

function hasAnyRootConflict(sessions: Map<string, string>, parents: Map<string, string>): boolean {
  const seen = new Map<string, string>()
  for (const [remoteKey, sessionID] of sessions) {
    const root = rootSession(parents, sessionID)
    if (root === "") continue
    const current = seen.get(root)
    if (current && current !== remoteKey) return true
    seen.set(root, remoteKey)
  }
  return false
}

function withParent(parents: Map<string, string>, sessionID: string, parentID: string): Map<string, string> {
  const next = new Map(parents)
  next.set(sessionID, parentID)
  return next
}

function queuePathWrite(path: string, write: () => Promise<void>): Promise<void> {
  const previous = writeQueues.get(path) ?? Promise.resolve()
  const run = previous.then(write)
  const next = run.catch(() => {})
  writeQueues.set(path, next)
  void next.then(() => {
    if (writeQueues.get(path) === next) writeQueues.delete(path)
  })
  return run
}

function remoteKeysForRoot(sessions: Map<string, string>, parents: Map<string, string>, root: string): string[] {
  if (root === "") return []
  const keys: string[] = []
  for (const [remoteKey, current] of sessions) {
    if (rootSession(parents, current) === root) keys.push(remoteKey)
  }
  return keys
}
